Parsers-V2: Refactor FastSAM & YOLO fixes.

kkeroo opened the pull request:
This PR refactors the last remaining parser FastSAM - because previously it was dependent on the YOLOExtended parser. Now its parent class is BaseParser. We dont need to pass input size anymore as it is calculated automatically from the output tensor shape by multiplying it with the stride.

I also moved two functions `_get_segmentation_outputs` and `reshape_seg_outputs` away from YOLO class into YOLO utils because the two functions are also used in FastSAM.

Validation logic in `ImgDetectionsExtended` for `masks` was not right for current masks in YOLOExtended. I just commented it because when we add new masks representetation we would probably need it.

`ImgDetectionsExtended` was not having option to add a confidence to the keypoint (YOLO Pose outputs also `visibility` for each keypoint - confidence in our case). So, this is also added.

YOLOExtedend parser was slightly modified. `build` method now correctly takes parameters from head config. Previously it was taking `metadata` from `head_config` but the `head_config` itself is acctually a `metadata`. And when using `KPTS_MODE` we extract also keypoint scores, previously they were given as `z` coordinate. @ptoupas can you check if this is correct because I didnt check with all YOLOs but I think this only affects YOLO Pose and we only have one, right?

jkbmrz reviewed (approved):
LGTM, just one docstring comment.

ptoupas commented:
@kkeroo I have also tested with `yolov8n-pose` and it works just fine with the changes that you've made to the keypoints.

kkeroo commented on depthai_nodes/ml/messages/creators/detection.py:
```diff
@@ -35,6 +36,8 @@ def create_detection_message(
     @type keypoints: Optional[np.array]
     @param masks: Masks of detected objects of shape (N, H, W).
     @type masks: Optional[np.ndarray]
+    @param keypoints_scores: Confidence scores of detected keypoints of shape (N,).
```
Fixed in [edaf738](https://github.com/luxonis/depthai-nodes/pull/109/commits/edaf738c4f4877891c97bebcd726f2ef0ba29e69).

kkeroo commented:
I moved mask related functions from yolo utils to mask utils. The functions are used in YOLO seg. and FastSAM. Added in [8c36e29](https://github.com/luxonis/depthai-nodes/pull/109/commits/8c36e29feb5d5084ed63b0c277f7fc02af822441).

In [417e0e0](https://github.com/luxonis/depthai-nodes/pull/109/commits/417e0e03c1d5378f8b336cdb334ecc03626dfd7a) I added `subtype` reading intead of `yolo_version` from `head_config` to be align with luxonis-ml schema @klemen1999.

klemen1999 commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -125,22 +133,35 @@ def build(
+        self.mask_conf = head_config.get("mask_conf", self.mask_conf)
+        self.anchors = head_config.get("anchors", self.anchors)
+        self.n_keypoints = head_config.get("n_keypoints", self.n_keypoints)
+        yolo_version = head_config.get("subtype", self.yolo_version)
```
I would rather go then with `subtype` instead of `yolo_version` on all names (attributes, variables, parameters, etc.) to be consistent. And Enum could be `YOLOSubtype` instead of `YOLOVersion`?

kkeroo replied:
Changed all names to `subtype` and `YOLOSubtype` in [009a00b](https://github.com/luxonis/depthai-nodes/pull/109/commits/009a00b8745a2cb13180796441ce3195d323e673).

klemen1999 reviewed (commented):
LGTM, just left some comments

kkeroo commented on depthai_nodes/ml/parsers/fastsam.py:
```diff
@@ -57,12 +62,18 @@ def __init__(
         point_label: Optional[int] = None,
         bbox: Optional[Tuple[int, int, int, int]] = None,
-    ):
+        yolo_outputs: List[str] = (
```
Yeah I know. I set that intentionally becuase pre-commit doesnt let you set mutable default arguments. The only option is to have it None.

kkeroo replied:
But then the problem arises if you dont use the `build` and you need to set them manually. I guess we still have arguments as None and then inside constructor set them to default values if the argument is None?

klemen1999 replied:
Yes let's do it like that: None and then set default in constructor if not passed directly. I believe we do it that way in other parsers as well right?

kkeroo replied:
Yes, added in [556b724](https://github.com/luxonis/depthai-nodes/pull/109/commits/556b724b816d31277703e0e8fa05029de6f9915c).

kkeroo commented on depthai_nodes/ml/parsers/fastsam.py:
```diff
@@ -83,58 +92,192 @@ def __init__(
+        return self
+
+    def setConfidenceThreshold(self, threshold: float) -> None:
+        """Sets the confidence score threshold for detected faces.
```
Copilot mistake :D fixed in [c62681c](https://github.com/luxonis/depthai-nodes/pull/109/commits/c62681cc2ad7eab5957d6599e7522d86fc2d53a0).

kkeroo commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -125,22 +131,35 @@ def build(
+        #     except ValueError as err:
+        #         raise ValueError(
+        #             f"Invalid YOLO version {metadata['yolo_version']}. Supported YOLO versions are {[e.value for e in YOLOVersion][:-1]}."
+        #         ) from err
```
Yes, fixed in [96ad40c](https://github.com/luxonis/depthai-nodes/pull/109/commits/96ad40c9f13cee8560494b25b75829b58ca3c1c7).

kkeroo commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -125,22 +133,35 @@ def build(
+            raise ValueError(
+                f"Invalid YOLO version {yolo_version}. Supported YOLO versions are {[e.value for e in YOLOVersion][:-1]}."
+            ) from err
+        # self.conf_threshold = metadata["conf_threshold"]
```
Yes, fixed in [96ad40c](https://github.com/luxonis/depthai-nodes/pull/109/commits/96ad40c9f13cee8560494b25b75829b58ca3c1c7).

kkeroo commented on depthai_nodes/ml/parsers/utils/yolo.py:
```diff
@@ -421,3 +422,34 @@ def decode_yolo_output(
+    protos_output = protos_output.transpose((0, 3, 1, 2))
+    protos_len = protos_output.shape[1]
+    masks_outputs_values = [o.transpose((0, 3, 1, 2)) for o in masks_outputs_values]
+    return protos_output, protos_len, masks_outputs_values
```
Added in [8c36e29](https://github.com/luxonis/depthai-nodes/pull/109/commits/8c36e29feb5d5084ed63b0c277f7fc02af822441).

klemen1999 commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -133,35 +133,20 @@ def build(
         except ValueError as err:
             raise ValueError(
-                f"Invalid YOLO version {yolo_version}. Supported YOLO versions are {[e.value for e in YOLOVersion][:-1]}."
+                f"Invalid YOLO version {subtype}. Supported YOLO versions are {[e.value for e in YOLOSubtype][:-1]}."
```
```suggestion
                f"Invalid YOLO subtype {subtype}. Supported YOLO subtypes are {[e.value for e in YOLOSubtype][:-1]}."
```

klemen1999 commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -212,17 +197,17 @@ def setAnchors(self, anchors: List[np.ndarray]) -> None:
 
-    def setYoloVersion(self, yolo_version: str) -> None:
+    def setSubtype(self, subtype: str) -> None:
         """Sets the version of the YOLO model.
```
```suggestion
        """Sets the subtype of the YOLO model.
```

klemen1999 commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -212,17 +197,17 @@ def setAnchors(self, anchors: List[np.ndarray]) -> None:
 
-        @param yolo_version: The version of the YOLO model.
-        @type yolo_version: YOLOVersion
+        @param subtype: The version of the YOLO model.
```
```suggestion
        @param subtype: The subtype of the YOLO model.
```

klemen1999 commented on depthai_nodes/ml/parsers/yolo.py:
```diff
@@ -212,17 +197,17 @@ def setAnchors(self, anchors: List[np.ndarray]) -> None:
         except ValueError as err:
             raise ValueError(
-                f"Invalid YOLO version {yolo_version}. Supported YOLO versions are {[e.value for e in YOLOVersion][:-1]}."
+                f"Invalid YOLO version {subtype}. Supported YOLO versions are {[e.value for e in YOLOSubtype][:-1]}."
```
```suggestion
                f"Invalid YOLO subtype {subtype}. Supported YOLO subtypes are {[e.value for e in YOLOSubtype][:-1]}."
```
